### PR TITLE
Fix/gpgnet

### PR DESCRIPTION
**Pull request by Sheeo**

### CODE REVIEW

File: lua/ui/lobby/lobby.lua [hunk 1]
```diff
@@ -2224,12 +2225,14 @@ function HostConvertPlayerToObserver(senderID, playerSlot, ignoreMsg)
         index = index + 1
     end
 
-    local playerName = gameInfo.PlayerOptions[playerSlot].PlayerName
```
Comment (ChrisKitching) [inline lua/ui/lobby/lobby.lua]: This variable is used later.

File: lua/ui/lobby/lobby.lua [hunk 2]
```diff
@@ -2224,12 +2225,14 @@ function HostConvertPlayerToObserver(senderID, playerSlot, ignoreMsg)
...
+    local ownerID = gameInfo.PlayerOptions[playerSlot].OwnerID
     gameInfo.Observers[index] = gameInfo.PlayerOptions[playerSlot]
     gameInfo.PlayerOptions[playerSlot] = nil
 
     if lobbyComm:IsHost() then
-        GpgNetSend('PlayerOption', string.format("team %s %d %s", playerName, -1, 0))
+        GpgNetSend('PlayerOption', ownerID, 'Team', -1)
+        GpgNetSend('PlayerOption', ownerID, 'Army', -1)
+        GpgNetSend('PlayerOption', ownerID, 'StartSpot', -index)
```
Comment (ChrisKitching) [inline lua/ui/lobby/lobby.lua]: Maybe have a neat SendPlayerOptions({Team = -1, Army = -1, StartSpot = -index}) function?

File: lua/ui/lobby/lobby.lua [hunk 3]
```diff
@@ -3835,7 +3837,6 @@ function InitLobbyComm(protocol, localPort, desiredPlayerName, localPlayerUID, n
         localPlayerID = myID
         localPlayerName = myName
 
-        GpgNetSend('connectedToHost', string.format("%d", hostID))
```
Comment (ChrisKitching) [inline lua/ui/lobby/lobby.lua]: This always struck me as a bit strange...

File: lua/ui/lobby/lobby.lua (hunk 1 above)
Comment (Sheeo) [inline lua/ui/lobby/lobby.lua]: Where?

File: lua/ui/lobby/lobby.lua [hunk 4]
```diff
@@ -3896,7 +3897,7 @@ function InitLobbyComm(protocol, localPort, desiredPlayerName, localPlayerUID, n
 
                 gameInfo.PlayerOptions[data.Slot][key] = val
                 if isHost then
-                    GpgNetSend('PlayerOption', data.Slot, key, val)
+                    GpgNetSend('PlayerOption', gameInfo.PlayerOptions[data.Slot].OwnerID, key, val)
```
Comment (ChrisKitching) [inline lua/ui/lobby/lobby.lua]: Maybe introduce a local variable to hold gameInfo.PlayerOptions[data.Slot], as it's used in a couple of places earlier too?

File: lua/ui/lobby/lobby.lua (hunk 2 above)
Comment (Sheeo) [inline lua/ui/lobby/lobby.lua]: I was thinking we would do this in a refactor anyway.

File: lua/ui/lobby/lobby.lua (hunk 3 above)
Comment (Sheeo) [inline lua/ui/lobby/lobby.lua]: Indeed it's nonsensical.

File: lua/ui/lobby/lobby.lua (hunk 4 above)
Comment (Sheeo) [inline lua/ui/lobby/lobby.lua]: Sure :)

Not the aim of this branch though

Comment (ChrisKitching) [inline lua/ui/lobby/lobby.lua]: Pfft. Clean up as you go along. There's enough shit everywhere. :P

**Comment by Sheeo** [comment]
Travis marks as failed because it tried pulling the defunct `refactor/gpgnet` branch over this one, for some reason (I changed my mind about the naming, since this has nothing to do with refactoring).

File: lua/ui/lobby/lobby.lua (hunk 4 above)
Comment (Sheeo) [inline lua/ui/lobby/lobby.lua]: lobbyComm.DataReceived as a whole needs a refactor, I think.

Introducing the local here would save us 1 table lookup, unless I do it at the toplevel of the function, where we are not guaranteed that PlayerOptions[data.Slot] has a value, so it becomes less meaningful.

**Comment by Sheeo** [comment]
Note that before we deploy this/the server update expecting this, the coop gamemode (And any other downgraded gamemode) has to be updated to be available with the current lobby.

**Comment by Sheeo** [comment]
Closing this. To be revived in a branch targeting whichever is to be released first: `release/3641` or `release/lobby-2.8`.

**Pull request closed without merging**